### PR TITLE
Fix mined stats and logs for forks

### PR DESCRIPTION
**Pull request by kernel32uiuc**
## Summary
There is too much confusion about whether we (as testnet users) have mined a block successfully or not. The reason is mostly about the logs. Ironfish logs the following message for the failed blocks as well. 
> _Successfully mined block 00000000076ef9860eba6ba1009946f9d6f3573e3f5ad2d0fdeb06177a6ade01 (15879) has 2 transactions_

miners:mined is also accurate, but it is very slow and occasionally hangs. That's why it is really important for users to easily understand whether they mine or not. 

This PR fixes this issue. When we log the 'Successfully mined block' message, it is now always accurate. 'status' command will also give the accurate stats for the current run.

Addressing #702.

## Breaking Change
```
[ ] Yes
```

### CODE REVIEW

File: ironfish/src/blockchain/blockchain.ts [hunk 1]
```diff
@@ -237,6 +237,7 @@ export class Blockchain {
 
     const result = await this.addBlock(genesis)
     Assert.isTrue(result.isAdded, `Could not seed genesis: ${result.reason || 'unknown'}`)
+    Assert.isTrue(result.isFork === false)
```
Comment (NullSoldier) [inline ironfish/src/blockchain/blockchain.ts:240]:
```suggestion
    Assert.isFalse(result.isFork)
```

Comment (kernel32uiuc) [reply ironfish/src/blockchain/blockchain.ts:240]: isFork can be null, so explicit false check is needed.

**Comment by sureshvenkata12** [comment]
@NullSoldier any issues with the PR?

**Comment by kernel32uiuc** [comment]
@NullSoldier @dguenther any updates on the PR?

**Comment by kernel32uiuc** [comment]
@NullSoldier I guess, this is not the direction you would like to proceed. Should I go ahead and abort this PR?

File: ironfish/src/mining/director.ts [hunk 2]
```diff
@@ -463,20 +463,22 @@ export class MiningDirector {
...
 
-    this.blocksMined++
+    if (!isFork) {
+      this.logger.info(
+        `Successfully mined block ${block.header.hash.toString('hex')} (${
+          block.header.sequence
+        }) has ${block.transactions.length} transactions`,
+      )
+
+      this.blocksMined++
```
Comment (NullSoldier) [inline ironfish/src/mining/director.ts:480]: Let's inverse this so it just returns and doesn't gossip a block on a fork.

File: ironfish/src/blockchain/blockchain.ts [hunk 3]
```diff
@@ -347,18 +351,18 @@ export class Blockchain {
...
     } catch (e) {
       if (e instanceof VerifyError) {
-        return { isAdded: false, reason: e.reason, score: e.score }
+        return { isAdded: false, isFork: null, reason: e.reason, score: e.score }
       }
       throw e
     }
 
-    return { isAdded: true, reason: null, score: null }
+    return { isAdded: true, isFork: isFork, reason: null, score: null }
```
Comment (NullSoldier) [inline ironfish/src/blockchain/blockchain.ts:365]:
```suggestion
    const isFork = !this.head.hash.equals(block.header)
    return { isAdded: true, isFork: isFork, reason: null, score: null }
```

File: ironfish/src/mining/director.ts (hunk 2 above)
Comment (cryptokur) [reply ironfish/src/mining/director.ts:480]: Do you mean adding a new MINED_RESULT such as `MINED_RESULT.FORK` and return that in case of `isFork`

Comment (kernel32uiuc) [reply ironfish/src/mining/director.ts:480]: something like that? 
`if (isFork) return MINED_RESULT.FORK`

File: ironfish/src/blockchain/blockchain.ts (hunk 3 above)
Comment (kernel32uiuc) [reply ironfish/src/blockchain/blockchain.ts:365]: sounds good. Why do we need a mutex lock here? Or did you just make a typo?

**Comment by kernel32uiuc** [comment]
@NullSoldier I addressed the feedback. Could you please review the changes?

**Comment by dguenther** [comment]
Yep, I'll take a look at this and try to get it through!

File: ironfish/src/blockchain/blockchain.ts (hunk 3 above)
Comment (NullSoldier) [reply ironfish/src/blockchain/blockchain.ts:366]: Two add blocks running at the same time can cause bugs.

**Review by NullSoldier — APPROVED** [review]
Directionally approved, derek is going to make small changes and to get this in.

**Comment by kernel32uiuc** [comment]
> Directionally approved, derek is going to make small changes and to get this in.

glad that this got merged. thanks for the help

**Pull request merged**